src/server/master_server/metadata_manager.cc

```diff
@@ -0,0 +1,102 @@
+#include "metadata_manager.h"
+#include <stack>
+
+namespace gfs {
+namespace server {
+
+MetadataManager::MetadataManager() {
+  lockManager_ = LockManager::GetInstance();
+}
+
+bool MetadataManager::CreateFileMetadata(const std::string& pathname) {
+  // Step 1. Lock (readerlock underneath) the parent directory first
+  ParentLocksAnchor parentLockAnchor(lockManager_, pathname);
+             
+  if (!parentLockAnchor.succ()) {
+    // If this operation fails, which means some of the parent directory does not
+    // exist, we return false
+    return false;
+  }
+
+  // Step 2. Add a new lock for this new file, and writeLock it
+  auto newLock(lockManager_->AddLockIfNonExist(pathname));
+  if (!newLock) {
+    // We still need to check whether the return is NULL because another thread
+    // could well successfully created a new lock for the same path
+    return false;
+  }
+   
+  absl::WriterMutexLock anchorForNewLock(newLock);
+  // Step 3. writeLock the global lock, instantiate a FileMetadata object
+  absl::WriterMutexLock anchorForGlobalLock(lockManager_->globalLock());
+  // The reason that we acquire the global lock is that we need to synchronization
+  // between write and read from the fileMetadata collection.
+  if(fileMetadata_.count(pathname)) {
+    return false;
+  }
+   
+  fileMetadata_[pathname] = std::make_shared<FileMetadata>();
+  return true;
+}
+
+bool MetadataManager::ExistFileMetadata(const std::string& pathname) const {
+  absl::ReaderMutexLock(lockManager_->globalLock());
+  return fileMetadata_.count(pathname);
+}
+
+std::shared_ptr<FileMetadata> MetadataManager::GetFileMetadata(
+  const std::string& pathname) const {
+  // readLock the global lock and retrieve the filemetadata. The reason for a
+  // readLock is because we are not mutating anything in the fileMetadata_.
+  absl::ReaderMutexLock(lockManager_->globalLock());
+  return fileMetadata_.at(pathname);
+}
+
+std::string MetadataManager::CreateChunkHandle(const std::string& pathname, 
+                                               uint32_t chunk_index) {
+  // Step 1. readlock the parent directories
+  ParentLocksAnchor parentLockAnchor(lockManager_, pathname);
+  if(!parentLockAnchor.succ()) {
+    // If this operation fails, which means some of the parent directory does not
+    // exist, we return false
+    return "";
+  }
+
+  // Step 2. readlock the global lock, fetch the data and unlock readerlock
+  auto fdata(GetFileMetadata(pathname));
```

Comment:
What you get is a shared pointer to the FileMetadata, so race condition and concurrent modification still apply if someone modifies this file metadata, as it contains other fields such as `maps`, because upon return of the function `GetFileMetadata` your reader lock is released.

Reply:
Thus, you should try to acquire the write lock on the path first, not latter.

Reply:
Another thread could modify `fdata` after this line, but that means the write lock must be acquired for it. That means the following lines starting from 70 gets blocked because you always have to acquire write lock when modifying this file metadata. 

GetFileMetadata acquires a lock that is for accessing the fileMetadata_ collection, after it is released it just means that someone else could add / remove FileMetadat* from this collection. The mutation of states in FileMetadata itself is protected by the GetLock returned by the LockManager. 

So I don't see this as a race condition. 

There is some interesting that can happen here still, namely at this line you fetch a pointer for FileMetadata, and what can happen next is that FileMetadata's pointer gets removed from the collection. But that is not unsafe because you can still mutate your data, it is just that the next time when you try to access the file metadata it is no longer there and you will find the returned chunk handle as deleted.